### PR TITLE
Continuation of `ndb` Property implementation (part 4)

### PR DESCRIPTION
**Pull request by dhermes**
This is still incomplete, it's a very large class.

In particular, this implements:

- `Property._apply_to_values`
- `Property._opt_call_to_base_type`
- `Property._call_from_base_type`
- `Property._opt_call_from_base_type`
- `Property._value_to_repr`
- `Property._get_user_value`
- `Property._get_base_value`
- `Property._get_base_value_unwrapped_as_list`

### CODE REVIEW

**Comment by dhermes** [comment]
@chrisrossi There are a few comments I'd like feedback on if you don't mind.

File: ndb/src/google/cloud/ndb/model.py [hunk 1]
```diff
@@ -852,6 +852,146 @@ def _retrieve_value(self, entity, default=None):
...
+        Raises:
+            ValueError: If the property is repeated but the base value is not
+                a :class:`list` or :data:`None`.
+            ValueError: If the property is not repeated but the base value is
+                not a :class:`_BaseValue`.
+        """
+        wrapped = self._get_base_value(entity)
+        if self._repeated:
+            if not isinstance(wrapped, list):
+                raise ValueError("Expected the wrapped quantity to be a list.")
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: If it were just me, I'd convert back to use an `assert` statement, since that's the best way to express, "Hey, I'm assumiing this will never happen, I want to test my assumption during development, and I want to express to future developers that I at least considered this edge case."

Some orgs hate `assert` statements, though.  I don't know if Google is one of them.

File: ndb/src/google/cloud/ndb/model.py [hunk 2]
```diff
@@ -852,6 +852,146 @@ def _retrieve_value(self, entity, default=None):
...
+            entity (Model): An entity to get a value from.
+
+        Returns:
+            Union[_BaseValue, List[_BaseValue]]: The original value
+            (if :class:`_BaseValue`) or the value converted to the base type
+            and wrapped.
+        """
+        return self._apply_to_values(entity, self._opt_call_to_base_type)
+
+    def _get_base_value_unwrapped_as_list(self, entity):
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: _get_base_values ?

File: ndb/src/google/cloud/ndb/model.py [hunk 3]
```diff
@@ -852,6 +852,146 @@ def _retrieve_value(self, entity, default=None):
...
+            missing or :data:`None` or empty, returns ``[]``.
+
+        Raises:
+            ValueError: If the property is repeated but the base value is not
+                a :class:`list` or :data:`None`.
+            ValueError: If the property is not repeated but the base value is
+                not a :class:`_BaseValue`.
+        """
+        wrapped = self._get_base_value(entity)
+        if self._repeated:
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: Sounds reasonable.

File: ndb/src/google/cloud/ndb/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: I don't know about Google, but I personally am not a fan of `assert` statements. They don't provide a useful error message to users running the code and they can be turned off if the right flag/env. var. is used, which is often a gotcha for devs and users.

[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: UPDATE: The Google Python [style guide][1] does not have any issue with `assert` (when used correctly).

[1]: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#244-decision

[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: Right. If you think there's any chance the condition would occur IRL, then the code should be left as-is. The `assert` is just a way to document assumptions being made about the code which should always be true, and provide a sanity check during development.  By the time code gets into production, turning them should be OK.  If you still think it could happen IRL, it shouldn't be an assert.  Deleting the check altogether, while fine (I wouldn't argue with you if that's what you eventually did), leaves the assumption implicit and opaque to future developers, whereas an `assert` tells future developers, explicitly, "I'm making this assumption."

File: ndb/src/google/cloud/ndb/model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/model.py]: I don't have strong feelings.  I just figured you were fishing for suggestions.  I don't mind the long name when it's used internally in this way.  I'd leave it as-is.

[the remark this answers is not in the document]
Comment (crwilcox) [reply ndb/src/google/cloud/ndb/model.py]: In agreement with @chrisrossi , I would ignore the TODO

File: ndb/src/google/cloud/ndb/model.py [hunk 4]
```diff
@@ -852,6 +852,146 @@ def _retrieve_value(self, entity, default=None):
...
+            if not isinstance(wrapped, list):
+                raise ValueError("Expected the wrapped quantity to be a list.")
+            return [w.b_val for w in wrapped]
+        else:
+            if wrapped is None:
+                return [None]
+            if not isinstance(wrapped, _BaseValue):
+                raise ValueError(
```
[the remark this answers is not in the document]
Comment (crwilcox) [reply ndb/src/google/cloud/ndb/model.py]: I think this could be removed. I do wonder if adding asserts would be useful or not.

File: ndb/src/google/cloud/ndb/model.py [hunk 5]
```diff
@@ -1026,6 +1166,42 @@ def call(value):
...
+        Returns:
+            Any: The transformed value store on the entity for this property.
+        """
+        value = self._retrieve_value(entity, self._default)
+        if self._repeated:
+            if value is None:
+                value = []
+                self._store_value(entity, value)
+            else:
+                # NOTE: This assumes, but does not check, that ``value`` is
```
Comment (crwilcox) [inline ndb/src/google/cloud/ndb/model.py]: Is this a verbatim paste from old code?

Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: No. I added it. No such comment existed before.

File: ndb/src/google/cloud/ndb/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply ndb/src/google/cloud/ndb/model.py]: SG @crwilcox. I'll remove it. I'm not a fan of `assert`s but reasonable people can disagree
